### 1) Add the Pandas dependency.

In [1]:
import pandas as pd

### 2)  Files to Load

In [2]:
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

### 3) Read the school data file and store it in a Pandas DataFrame

In [3]:
school_data_df = pd.read_csv(school_data_to_load)
school_data_df

School ID            school_name      type  size   budget
0           0      Huang High School  District  2917  1910635
1           1   Figueroa High School  District  2949  1884411
2           2    Shelton High School   Charter  1761  1056600
3           3  Hernandez High School  District  4635  3022020
4           4    Griffin High School   Charter  1468   917500
5           5     Wilson High School   Charter  2283  1319574
6           6    Cabrera High School   Charter  1858  1081356
7           7     Bailey High School  District  4976  3124928
8           8     Holden High School   Charter   427   248087
9           9       Pena High School   Charter   962   585858
10         10     Wright High School   Charter  1800  1049400
11         11  Rodriguez High School  District  3999  2547363
12         12    Johnson High School  District  4761  3094650
13         13       Ford High School  District  2739  1763916
14         14     Thomas High School   Charter  1635  1043130

### 4) Read the student data file and store it in a Pandas DataFrame.

In [4]:
student_data_df = pd.read_csv(student_data_to_load,  index_col=[0])
student_data_df.head()

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  
0          79  
1          61  
2          60  
3          58  
4          84

### 5) Determine if there are any missing values in the school data - using count() method
We can see there are no missing values because there are 15 rows that contain data in schools_complete.csv.

In [5]:
school_data_df.count()

School ID      15
school_name    15
type           15
size           15
budget         15
dtype: int64

### 6) Determine if there are any missing values in the student data -  using count() method
Again, no missing values b/c output shows 39170 rows for the csv file

In [6]:
student_data_df.count()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

### 7) Determine if there are any missing values in the school data - using isnull() method
When we execute this code, we see that every row that is not empty is given the Boolean value "False," which tells us that there are no missing values.

In [7]:
school_data_df.isnull()

School ID  school_name   type   size  budget
0       False        False  False  False   False
1       False        False  False  False   False
2       False        False  False  False   False
3       False        False  False  False   False
4       False        False  False  False   False
5       False        False  False  False   False
6       False        False  False  False   False
7       False        False  False  False   False
8       False        False  False  False   False
9       False        False  False  False   False
10      False        False  False  False   False
11      False        False  False  False   False
12      False        False  False  False   False
13      False        False  False  False   False
14      False        False  False  False   False

### 8) Determine if there are any missing values in the student data - using isnull() method
A) Output shows that there are no rows that contain missing values; they are all "False." 

In [8]:
student_data_df.isnull()

Student ID  student_name  gender  grade  school_name  reading_score  \
0           False         False   False  False        False          False   
1           False         False   False  False        False          False   
2           False         False   False  False        False          False   
3           False         False   False  False        False          False   
4           False         False   False  False        False          False   
...           ...           ...     ...    ...          ...            ...   
39165       False         False   False  False        False          False   
39166       False         False   False  False        False          False   
39167       False         False   False  False        False          False   
39168       False         False   False  False        False          False   
39169       False         False   False  False        False          False   

       math_score  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
39165       False  
39166       False  
39167       False  
39168       False  
39169       False  

[39170 rows x 7 columns]

B) Determine if there are any missing values in the student data using -isnull() more efficiently 

 - Chaining two (+) methods or functions together separated by period

 - Get the total number of empty rows, or rows that are "True" 

In [9]:
student_data_df.isnull().sum()

Student ID       0
student_name     0
gender           0
grade            0
school_name      0
reading_score    0
math_score       0
dtype: int64

### 9) Determine if there are not any missing values in the school data - using notnull() method
Output returns a copy of our school_data_df DataFrame, where all the rows that do not have any missing values are labeled "True." 
 - Returns the opposite output of the isnull() method

In [10]:
school_data_df.notnull()

School ID  school_name  type  size  budget
0        True         True  True  True    True
1        True         True  True  True    True
2        True         True  True  True    True
3        True         True  True  True    True
4        True         True  True  True    True
5        True         True  True  True    True
6        True         True  True  True    True
7        True         True  True  True    True
8        True         True  True  True    True
9        True         True  True  True    True
10       True         True  True  True    True
11       True         True  True  True    True
12       True         True  True  True    True
13       True         True  True  True    True
14       True         True  True  True    True

### 10) Determine if there are not any missing values in the student data - using notnull() method
Chaining two (+) methods or functions together separated by period
 - Get sum of all columns that are "True" (not null)

In [11]:
student_data_df.notnull().sum()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

### 11) Determine the data types for the school DataFrame -- Using dtypes
"School ID," "size," and "budget" are integers, which is the correct data type for performing arithmetic calculations.

In [12]:
school_data_df.dtypes

School ID       int64
school_name    object
type           object
size            int64
budget          int64
dtype: object

### 12) Determine the data types for the budget column -- using dtype
Use df.column.dtype if the column name doesn't have spaces in the name, like school_name. 
* If there are spaces, we must use df["column"].dtype like this: school_data_df["budget"].dtype

In [13]:
school_data_df.budget.dtype

dtype('int64')

### 13) Determine data types for the student_data DataFrame -- Using dtypes
Returns each column name and the data type of each column. 
* "Student ID," "reading_score," and "math_score" are integers, which is the correct data type for performing arithmetic calculations

In [14]:
student_data_df.dtypes

Student ID        int64
student_name     object
gender           object
grade            object
school_name      object
reading_score     int64
math_score        int64
dtype: object

### 14) Fix Students2019 Names
Using the same code that we tested for cleaning the students' names, we will fix the students' names so they match their school records

In [15]:
# a) Add each prefix and suffix to remove to a list.
prefixes_suffixes = ["Dr. ", "Mr. ","Ms. ", "Mrs. ", "Miss ", 
                     " MD", " DDS", " DVM", " PhD"]

# b) Iterate through words (pre/suffxs) in the "prefixes_suffixes" list 
for word in prefixes_suffixes:
    
    # c) Replace words in the list (pre/suffxs) with an empty space, "".
    student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")

# d) Print student DataFrame
student_data_df.head(10)

C:\Users\ewaas\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:9: FutureWarning: The default value of regex will change from True to False in a future version.
  if __name__ == '__main__':


Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   
7           7     Nicole Baker      F  12th  Huang High School             96   
8           8     Michael Roth      M  10th  Huang High School             95   
9           9   Matthew Greene      M  10th  Huang High School             96   

   math_score  
0          79  
1          61  
2          60  
3          58  
4          84  
5          94  
6          80  
7          69  
8          87  
9          84

# Save data back to CSV File 
(apparently it did that when I saved the cleaning_student_names file)

### 15) Merge dataFrames on shared columns using merge()
Combine the two dataFrames into a single dataset.
* School ID, type, size, and budget should be the last columns now

New DF is results of merging DF#2 ("right" DF = school_data_df) with DF#1 ("Left" DF = student_data_df) 
* L and R DFs reflect the order in which they appear inside parentheses.

* Can also write the column name "on="school_name" just once

In [16]:
# a) Create new dataFrame for the merged df's (school_data_complete_df)  
# b)  Use parameter "on=[]" , which is equal to a list of columns that 
        # are identicalfrom each DF. (school_name in this case) 
school_data_complete_df = pd.merge(student_data_df, school_data_df, on=["school_name", "school_name"])

# c) Print out the merged DF
school_data_complete_df.head()

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
3          58          0  District  2917  1910635  
4          84          0  District  2917  1910635

### 16) Get the total number of students in the school district
We are counting all the items in each row for each column that is not null  

Use this code to get the total number of rows (or values) for each column: 
student_count = school_data_complete_df.count()
* Output: list of each column name with 39170 rows (or values in each column)


Since they are all the same, we can use any column to get the student count. 

Use this modified format to find the student count for a particular column: 
school_data_complete_df[column].count()

We are going to assign the "student_count" variable to a column that identifies with students to get the student count. 
* We picked Student ID.

Output should change from list with each column displaying 39170 to just 39170
* We can't use this because it just gives us 39170

In [17]:
student_count = school_data_complete_df["Student ID"].count()
student_count

39170

### 17) Calculate the total number of schools

#### A) Calculate the total number of schools using count()
Use school_data_df DataFrame and assign a variabel to one of the columns in the DF. 
* This gives a count of how many school names there are in the school data DF
* Refer to #3. There are 15 rows in that data frame, one for each school. 

In [18]:
school_count = school_data_df["school_name"].count()
school_count

15

#### B) Calculate the total number of schools using .unique()
* Returns an ndarray = n-dimensional array of all the unique values of that column

In [19]:
school_count_2 = school_data_complete_df["school_name"].unique()
school_count_2

array(['Huang High School', 'Figueroa High School', 'Shelton High School',
       'Hernandez High School', 'Griffin High School',
       'Wilson High School', 'Cabrera High School', 'Bailey High School',
       'Holden High School', 'Pena High School', 'Wright High School',
       'Rodriguez High School', 'Johnson High School', 'Ford High School',
       'Thomas High School'], dtype=object)

### 18) Get the total budget

#### A) Refer to school_data_complete_df in step 15. 
##### (The dataFrame we created by erging student_data_df with school_data_df)
In this DF, each student is associated with a school, and all data related to that school
* If we use, sum() on the budget column on this DF, we get: (first row of code in cell below)

Output: 82932329558
* This summed all 39,170 rows in the budget column of the df. 
* This is incorrect. 

In [20]:
total_budget_0 = school_data_complete_df["budget"].sum()
total_budget_0

82932329558

#### B) To calculate the total budget: 
Apply the .sum()method to budget column in school_data_df 
* Refer to #3
Output should be 24649428. 
* Verify by using sum(E3:E16) in csv file. Don't save. It matches.

In [21]:
total_budget = school_data_df["budget"].sum()
total_budget

24649428

### 19) Get the test score averages

#### A) Calculate the average reading score
* Create a variable (avg reading score)
* Find mean of all values in "reading score" column within school_data_complete_df
* Print the average reading score for all students.

In [22]:
average_reading_score = school_data_complete_df["reading_score"].mean()
average_reading_score

81.87784018381414

#### B) Calculate the average math score

* Create a variable (avg math score)
* Find mean of all values in "math score" column within school_data_complete_df
* Print the average reading score for all students.

In [23]:
average_math_score = school_data_complete_df["math_score"].mean()
average_math_score

78.98537145774827

### 20) Get the Passing Percentages

#### A) Determine the passing grade
Create two new variable to store the passing grades for math and reading 
* (passing_math and passing_reading)

Within the math/reading_score column of school_data_complete_df, determine if value is greater than or equal to 70, the value we determine to be passing.

In [24]:
passing_math = school_data_complete_df["math_score"]>=70
passing_reading = school_data_complete_df["reading_score"]>=70

#### B) Print Math
Output: True >= 70 ; False < 70

In [25]:
passing_math.head()

0     True
1    False
2    False
3    False
4     True
Name: math_score, dtype: bool

#### C) Print Reading
Output: True >= 70 ; False < 70

In [26]:
passing_reading.head()

0    False
1     True
2     True
3    False
4     True
Name: reading_score, dtype: bool

### 21) Get the Number of Students Who Passed Math / Reading

#### A) Get Number of Students Who Passed Math into a new dataFrame
* Create new variable to store students who are passing math (passing_math)
* Filter school_data_complete_df for only "True" cases
* Add [school_data_complete_df["math_score"] >= 70] within [brackets] to filter through students in the school_data_complete_df who had a math score >=70
* Print list of students who passed math 

In [27]:
passing_math = school_data_complete_df[school_data_complete_df["math_score"] >= 70]
passing_math.head()

Student ID   student_name gender grade        school_name  reading_score  \
0           0   Paul Bradley      M   9th  Huang High School             66   
4           4     Bonnie Ray      F   9th  Huang High School             97   
5           5  Bryan Miranda      M   9th  Huang High School             94   
6           6  Sheena Carter      F  11th  Huang High School             82   
8           8   Michael Roth      M  10th  Huang High School             95   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635

#### B) Get Number of Students Who Passed Reading into a new dataFrame (Repurpose Code for Math)
* Create new variable to store students who are passing reading (passing_reading)
* Filter school_data_complete_df for only "True" cases
* Add [school_data_complete_df["reading_score"] >= 70] within [brackets] to filter through students in the school_data_complete_df who had a reading score >=70
* Print list of students who passed reading

In [28]:
passing_reading = school_data_complete_df[school_data_complete_df["reading_score"] >= 70]
passing_reading.head()

Student ID     student_name gender grade        school_name  reading_score  \
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   

   math_score  School ID      type  size   budget  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635

#### C) Get the number of students who passed math / reading from the new dataFrames we just created in Step 21
* Calculate the number of students passing math.
* Calculate the number of students passing reading.

* Print the lists with the counts of students who passed for math/reading

In [29]:
passing_math_count = passing_math["student_name"].count()
passing_reading_count = passing_reading["student_name"].count()

In [30]:
print(passing_math_count)
print(passing_reading_count)

29370
33610


### 22) Get the Percentage of Students Who Passed Math/Reading
Calculate percentage that passed Math

Calculate the percent that passed Reading
* Recall that student count was calculated using: school_data_complete_df["Student ID"].count; 
* Gave us a total of 39170 students

Print the lists with the percentages of students who passed for math/reading

In [31]:
passing_math_percentage = passing_math_count / float(student_count) * 100
passing_reading_percentage = passing_reading_count / float(student_count) * 100

In [32]:
print(passing_math_percentage)
print(passing_reading_percentage)

74.9808526933878
85.80546336482001


### 23) Calculate Overall Passing Percentage (if student passed BOTH math and reading)
Calculate percentage of students who passed BOTH math AND reading
* Join by using "&"

Print list of students who passed BOTH math and reading

In [33]:
passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & 
                                               (school_data_complete_df["reading_score"] >= 70)]
passing_math_reading.head()

Student ID    student_name gender grade        school_name  reading_score  \
4           4      Bonnie Ray      F   9th  Huang High School             97   
5           5   Bryan Miranda      M   9th  Huang High School             94   
6           6   Sheena Carter      F  11th  Huang High School             82   
8           8    Michael Roth      M  10th  Huang High School             95   
9           9  Matthew Greene      M  10th  Huang High School             96   

   math_score  School ID      type  size   budget  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635  
9          84          0  District  2917  1910635

### 24) Get Total Count and Percentage of students who passed BOTH math and reading
Calculate the number of students who passed both math and reading.

Calculate the overall passing percentage.

In [34]:
overall_passing_math_reading_count = passing_math_reading["student_name"].count()
overall_passing_math_reading_count

25528

In [35]:
overall_passing_percentage = overall_passing_math_reading_count / student_count * 100
overall_passing_percentage

65.17232575950983

### 25) Create a New dataFrame with columns for all the metrics we calculated 
Adding a list of values with keys to create a new DataFrame.

Create a list of dictionaries, where keys = column names and values = calculated metrics

We did this bc DF has no indexing; lists have natural indexing

In [36]:
district_summary_df = pd.DataFrame(
          [{"Total Schools": school_count,
          "Total Students": student_count,
          "Total Budget": total_budget,
          "Average Math Score": average_math_score,
          "Average Reading Score": average_reading_score,
          "% Passing Math": passing_math_percentage,
         "% Passing Reading": passing_reading_percentage,
        "% Overall Passing": overall_passing_percentage}])

district_summary_df

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

### 26) Write and use a function to get the percentage of students who passed math
We know the number of students that passed math 
* pass_math_count
And the total number of students for a school district
* student_count

A) Define a function that calculates the percentage of students that passed both math and reading and prints the passing percentage to the output when the function is called.

In [37]:
def passing_math_percent(pass_math_count, student_count):
    return pass_math_count / float(student_count) * 100

B) Assign the passing_math_count and the total_student_count variables 29,730 and 39,170, respectively
* When we run this cell, nothing will happen until we call the function.

In [38]:
passing_math_count = 29370
total_student_count = 39170

C) Now we can call the function in a new cell with the two values

In [39]:
passing_math_percent(passing_math_count, total_student_count)

74.9808526933878

### 27) Format the "Total Students" to have the comma for a thousands separator.
Let's apply this formatting technique to format the Total Students column with a thousands separator using the basic syntax: df["column"] = df["column"].map("{:,}".format)

Now let's replace df["column"] with district_summary_df["Total Students"], so our code looks like the following:

In [40]:
district_summary_df["Total Students"] = district_summary_df["Total Students"].map("{:,}".format)
district_summary_df["Total Students"]

0    39,170
Name: Total Students, dtype: object

### 28) Format "Total Budget" to have the comma for a thousands separator, a decimal separator, and a $

In [41]:
district_summary_df["Total Budget"] = district_summary_df["Total Budget"].map("${:,.2f}".format)
district_summary_df["Total Budget"]

0    $24,649,428.00
Name: Total Budget, dtype: object

### 29) We need to format the remaining columns accordingly:
* The "Average Reading Score" column will be formatted to one decimal place.
* The "Average Math Score" column will be formatted to one decimal place.
* The "% Passing Reading" column will be formatted to the nearest whole number percentage.
* The "% Passing Math" column will be formatted to the nearest whole number percentage
* The "% Overall Passing" column will be formatted to the nearest whole number percentage.

In [42]:
district_summary_df["Average Math Score"] = district_summary_df["Average Math Score"].map("{:.1f}".format)

district_summary_df["Average Reading Score"] = district_summary_df["Average Reading Score"].map("{:.1f}".format)

district_summary_df["% Passing Math"] = district_summary_df["% Passing Math"].map("{:.0f}".format)

district_summary_df["% Passing Reading"] = district_summary_df["% Passing Reading"].map("{:.0f}".format)

district_summary_df["% Overall Passing"] = district_summary_df["% Overall Passing"].map("{:.0f}".format)

Verify changes in district_summary_df

In [43]:
district_summary_df

Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00               79.0   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                  81.9             75                86                65

### 30) How to Reorder Columns if they are not in the Right Order 

Here is the standard format for reordering columns:

* Reorder the columns in the order you want them to appear.

new_column_order = ["column2", "column4", "column1"]


* Assign a new or the same DataFrame the new column order.

df = df[new_column_order]


Here is the exact code to reorder this set of data: 
* Reorder the columns in the order you want them to appear.

new_column_order = ["Total Schools", "Total Students", "Total Budget","Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

* Assign district summary df the new column order.

district_summary_df = district_summary_df[new_column_order]
district_summary_df

## Generate the School Summary (#31 - # )

### 31) Set the Index to the School Name

* Create a new DataFrame - school summary
* As mentioned, the index needs to be the school_name with "School Type" as the first column. 

* We can use the data from the school_data_df DataFrame because the "school_name" and "type" columns are both Series.
* We'll take these Series and use one as the index and the other as the first column. 

#### A) Determine School Type

In [44]:
per_school_types = school_data_df.set_index(["school_name"])["type"]
per_school_types

school_name
Huang High School        District
Figueroa High School     District
Shelton High School       Charter
Hernandez High School    District
Griffin High School       Charter
Wilson High School        Charter
Cabrera High School       Charter
Bailey High School       District
Holden High School        Charter
Pena High School          Charter
Wright High School        Charter
Rodriguez High School    District
Johnson High School      District
Ford High School         District
Thomas High School        Charter
Name: type, dtype: object

#### B) Add the per_school_types into a DataFrame for testing.

In [45]:
df = pd.DataFrame(per_school_types)
df

type
school_name                    
Huang High School      District
Figueroa High School   District
Shelton High School     Charter
Hernandez High School  District
Griffin High School     Charter
Wilson High School      Charter
Cabrera High School     Charter
Bailey High School     District
Holden High School      Charter
Pena High School        Charter
Wright High School      Charter
Rodriguez High School  District
Johnson High School    District
Ford High School       District
Thomas High School      Charter

### 32) Get the Student Count Per School

#### A) Calculate the total student count (from the school_data_df)
There are two DataFrames that have the total number of students per school: school_data_df and school_data_complete_df. 
* First, we will use the school_data_df DataFrame.

Looking at the school_data_df DataFrame, we can see the student count is in the "size" column
* Let's get the student count in the school_data_df DataFrame


Output: Unfortunately, this Series doesn't have an index with "school_name."
* Therefore, we can't use the "size" column from school_data_df to get the count of the student population.

In [46]:
per_school_counts = school_data_df["size"]
per_school_counts

0     2917
1     2949
2     1761
3     4635
4     1468
5     2283
6     1858
7     4976
8      427
9      962
10    1800
11    3999
12    4761
13    2739
14    1635
Name: size, dtype: int64

#### B)  Make school name the index
To fix this, (Output in Part A doesn't use "school_name" as the index), we can use the set_index() method on the "school_name" column, and then select the "size" column to display the student count for each school.

In [47]:
per_school_counts = school_data_df.set_index(["school_name"])["size"]
per_school_counts

school_name
Huang High School        2917
Figueroa High School     2949
Shelton High School      1761
Hernandez High School    4635
Griffin High School      1468
Wilson High School       2283
Cabrera High School      1858
Bailey High School       4976
Holden High School        427
Pena High School          962
Wright High School       1800
Rodriguez High School    3999
Johnson High School      4761
Ford High School         2739
Thomas High School       1635
Name: size, dtype: int64

#### C) Calculate the total student count per high school (from the school_data_complete_df)
Count the number of times a high school appears using value_counts() on the "school_name" column

The value_counts() method will return a Series of data with the number of times each school_name appears in a row. 

As long as we make sure the index is the "school_name" column, we can use either method.

In [48]:
per_school_counts = school_data_complete_df["school_name"].value_counts()
per_school_counts

Bailey High School       4976
Johnson High School      4761
Hernandez High School    4635
Rodriguez High School    3999
Figueroa High School     2949
Huang High School        2917
Ford High School         2739
Wilson High School       2283
Cabrera High School      1858
Wright High School       1800
Shelton High School      1761
Thomas High School       1635
Griffin High School      1468
Pena High School          962
Holden High School        427
Name: school_name, dtype: int64

### 33) Get the Budget Per Student
In order to find the budget per student for each school, we need to divide the budget for each school by the number of students at each school

#### A) Calculate the Total School Budget 
* If we look at the school_data_df DataFrame, The budget for each school is listed in the budget column

* The data type for the budget column is int64, which is suitable for calculations that we need to perform in order to find the budget per student

In [49]:
per_school_budget = school_data_df.set_index(["school_name"])["budget"]
per_school_budget

school_name
Huang High School        1910635
Figueroa High School     1884411
Shelton High School      1056600
Hernandez High School    3022020
Griffin High School       917500
Wilson High School       1319574
Cabrera High School      1081356
Bailey High School       3124928
Holden High School        248087
Pena High School          585858
Wright High School       1049400
Rodriguez High School    2547363
Johnson High School      3094650
Ford High School         1763916
Thomas High School       1043130
Name: budget, dtype: int64

#### B) Calculate the per Capita Spending (Budget per Student)
Divide the per_school_budget by the per_school_counts

* We can perform this calculation because the per_school_budget and per_school_counts are Series, both data types are int64, and both have the same index.

* When we execute this code, we get a Series with the school name as the index with a column showing the budget per student

In [50]:
per_school_capita = per_school_budget / per_school_counts
per_school_capita

Bailey High School       628.0
Cabrera High School      582.0
Figueroa High School     639.0
Ford High School         644.0
Griffin High School      625.0
Hernandez High School    652.0
Holden High School       581.0
Huang High School        655.0
Johnson High School      650.0
Pena High School         609.0
Rodriguez High School    637.0
Shelton High School      600.0
Thomas High School       638.0
Wilson High School       578.0
Wright High School       583.0
dtype: float64

# 34) Get the Score Averages Per School

We've used the set_index() method on the school_name column in student_data_df to get data from another column, just like how we retrieved the school budget using school_data_df.set_index(["school_name"])["budget"].

* Reuse this line of code, replacing budget with math_score. 


Output: 
Unfortunately, we can't use the school_data_df DataFrame, as there aren't any columns containing grades. 
* We also can't use the set_index() method on the school_name column in student_data_df because there are too many occurrences of the school_name column.


#### A) Calculate the Math Scores

In [51]:
student_school_math = student_data_df.set_index(["school_name"])["math_score"]
student_school_math

school_name
Huang High School     79
Huang High School     61
Huang High School     60
Huang High School     58
Huang High School     84
                      ..
Thomas High School    90
Thomas High School    70
Thomas High School    84
Thomas High School    90
Thomas High School    75
Name: math_score, Length: 39170, dtype: int64

#### B) Calculate Math Scores Using groupby()

The groupby() function will split an object (like a DataFrame), apply a mathematical operation, and combine the results. 
* This can be used to group large amounts of data when we want to compute mathematical operations on these groups.

* The mathematical operation we will apply to the groupby() function is the mean() method

In [52]:
per_school_averages = school_data_complete_df.groupby(["school_name"]).mean()
per_school_averages

Student ID  reading_score  math_score  School ID  \
school_name                                                               
Bailey High School        20358.5      81.033963   77.048432        7.0   
Cabrera High School       16941.5      83.975780   83.061895        6.0   
Figueroa High School       4391.0      81.158020   76.711767        1.0   
Ford High School          36165.0      80.746258   77.102592       13.0   
Griffin High School       12995.5      83.816757   83.351499        4.0   
Hernandez High School      9944.0      80.934412   77.289752        3.0   
Holden High School        23060.0      83.814988   83.803279        8.0   
Huang High School          1458.0      81.182722   76.629414        0.0   
Johnson High School       32415.0      80.966394   77.072464       12.0   
Pena High School          23754.5      84.044699   83.839917        9.0   
Rodriguez High School     28035.0      80.744686   76.842711       11.0   
Shelton High School        6746.0      83.725724   83.359455        2.0   
Thomas High School        38352.0      83.848930   83.418349       14.0   
Wilson High School        14871.0      83.989488   83.274201        5.0   
Wright High School        25135.5      83.955000   83.682222       10.0   

                         size     budget  
school_name                               
Bailey High School     4976.0  3124928.0  
Cabrera High School    1858.0  1081356.0  
Figueroa High School   2949.0  1884411.0  
Ford High School       2739.0  1763916.0  
Griffin High School    1468.0   917500.0  
Hernandez High School  4635.0  3022020.0  
Holden High School      427.0   248087.0  
Huang High School      2917.0  1910635.0  
Johnson High School    4761.0  3094650.0  
Pena High School        962.0   585858.0  
Rodriguez High School  3999.0  2547363.0  
Shelton High School    1761.0  1056600.0  
Thomas High School     1635.0  1043130.0  
Wilson High School     2283.0  1319574.0  
Wright High School     1800.0  1049400.0

#### C)  Get the average math score and reading score for each school
We don't want all of the data from output B) in the school summary DataFrame, just the reading and math scores
* Add the math_score and reading_score columns at the end

In [53]:
per_school_math = school_data_complete_df.groupby(["school_name"]).mean()["math_score"]
per_school_reading = school_data_complete_df.groupby(["school_name"]).mean()["reading_score"]
print(per_school_math)
print(per_school_reading)

school_name
Bailey High School       77.048432
Cabrera High School      83.061895
Figueroa High School     76.711767
Ford High School         77.102592
Griffin High School      83.351499
Hernandez High School    77.289752
Holden High School       83.803279
Huang High School        76.629414
Johnson High School      77.072464
Pena High School         83.839917
Rodriguez High School    76.842711
Shelton High School      83.359455
Thomas High School       83.418349
Wilson High School       83.274201
Wright High School       83.682222
Name: math_score, dtype: float64
school_name
Bailey High School       81.033963
Cabrera High School      83.975780
Figueroa High School     81.158020
Ford High School         80.746258
Griffin High School      83.816757
Hernandez High School    80.934412
Holden High School       83.814988
Huang High School        81.182722
Johnson High School      80.966394
Pena High School         84.044699
Rodriguez High School    80.744686
Shelton High School      83.72572

# 35) Get the Passing Percentages per School
To get the passing percentages, we need to: (Modify the code for this step, same structure)
A) Determine what is the passing grade. (Step 20)
* passing_math = school_data_complete_df["math_score"]>=70
* passing_reading = school_data_complete_df["reading_score"]>=70

B) Get the number of students who passed math OR reading. (Step 21 A & B)
* passing_math = school_data_complete_df[school_data_complete_df["math_score"] >= 70]
* passing_reading = school_data_complete_df[school_data_complete_df["reading_score"] >= 70]

C) Get the students who passed BOTH math AND passed reading (Step 23)
* passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & 
                                               (school_data_complete_df["reading_score"] >= 70)]

#### A) Calculate the passing scores by creating a filtered DataFrame.
So variables don't get reassigned, we'll use the same code but assign each calculation to new variables that reflect the students who passed math or reading for each school

Output: However, we need to get the average reading and math scores for each school. This dataframe uses student id (or not index?). We need to make school_name the index. 

In [54]:
per_school_passing_math = school_data_complete_df[(school_data_complete_df["math_score"] >= 70)]
per_school_passing_reading = school_data_complete_df[(school_data_complete_df["reading_score"] >= 70)]

In [55]:
per_school_passing_math.head()

Student ID   student_name gender grade        school_name  reading_score  \
0           0   Paul Bradley      M   9th  Huang High School             66   
4           4     Bonnie Ray      F   9th  Huang High School             97   
5           5  Bryan Miranda      M   9th  Huang High School             94   
6           6  Sheena Carter      F  11th  Huang High School             82   
8           8   Michael Roth      M  10th  Huang High School             95   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635

In [56]:
per_school_passing_reading.head()

Student ID     student_name gender grade        school_name  reading_score  \
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   

   math_score  School ID      type  size   budget  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635

#### B) Make school_name the index for this dataset. 
* Index needs to be the school_name, and we need to get the number of students in the per_school_passing_math and the per_school_passing_reading DataFrames.

QUESTION: From the per_school_passing_math and the per_school_passing_reading DataFrames, how can we get the index to be the school_name? 
ANSWER:  Use groupby(["school_name"])

To create a Series that has school_name as the index, we'll use groupby(["school_name"]) on the per_school_passing_math and the per_school_passing_reading DataFrames


Output: When we print the per_school_passing_math/reading, the output will be a Series where the index is the school name and the column is the number of students who passed math for each school

In [57]:
per_school_passing_math = per_school_passing_math.groupby(["school_name"]).count()["student_name"]
per_school_passing_reading = per_school_passing_reading.groupby(["school_name"]).count()["student_name"]

In [58]:
print(per_school_passing_math)
print(per_school_passing_reading)

school_name
Bailey High School       3318
Cabrera High School      1749
Figueroa High School     1946
Ford High School         1871
Griffin High School      1371
Hernandez High School    3094
Holden High School        395
Huang High School        1916
Johnson High School      3145
Pena High School          910
Rodriguez High School    2654
Shelton High School      1653
Thomas High School       1525
Wilson High School       2143
Wright High School       1680
Name: student_name, dtype: int64
school_name
Bailey High School       4077
Cabrera High School      1803
Figueroa High School     2381
Ford High School         2172
Griffin High School      1426
Hernandez High School    3748
Holden High School        411
Huang High School        2372
Johnson High School      3867
Pena High School          923
Rodriguez High School    3208
Shelton High School      1688
Thomas High School       1591
Wilson High School       2204
Wright High School       1739
Name: student_name, dtype: int64


### Determine the Percentage of Students Passing Math and Reading
Calculate the percentage of passing math and reading scores per school.

In [59]:
per_school_passing_math = per_school_passing_math / per_school_counts * 100
per_school_passing_reading = per_school_passing_reading / per_school_counts * 100
per_school_passing_math

Bailey High School       66.680064
Cabrera High School      94.133477
Figueroa High School     65.988471
Ford High School         68.309602
Griffin High School      93.392371
Hernandez High School    66.752967
Holden High School       92.505855
Huang High School        65.683922
Johnson High School      66.057551
Pena High School         94.594595
Rodriguez High School    66.366592
Shelton High School      93.867121
Thomas High School       93.272171
Wilson High School       93.867718
Wright High School       93.333333
dtype: float64

In [60]:
per_school_passing_reading

Bailey High School       81.933280
Cabrera High School      97.039828
Figueroa High School     80.739234
Ford High School         79.299014
Griffin High School      97.138965
Hernandez High School    80.862999
Holden High School       96.252927
Huang High School        81.316421
Johnson High School      81.222432
Pena High School         95.945946
Rodriguez High School    80.220055
Shelton High School      95.854628
Thomas High School       97.308869
Wilson High School       96.539641
Wright High School       96.611111
dtype: float64

#### C) Get the Overall Passing Percentage for All Students for Each School

In [61]:
per_passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]
per_passing_math_reading.head()

Student ID    student_name gender grade        school_name  reading_score  \
4           4      Bonnie Ray      F   9th  Huang High School             97   
5           5   Bryan Miranda      M   9th  Huang High School             94   
6           6   Sheena Carter      F  11th  Huang High School             82   
8           8    Michael Roth      M  10th  Huang High School             95   
9           9  Matthew Greene      M  10th  Huang High School             96   

   math_score  School ID      type  size   budget  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635  
9          84          0  District  2917  1910635

#### D) Get the total number of students from EACH SCHOOL who passed BOTH math AND reading from the per_passing_math_reading DataFrame
To do this, we need to set the index as school_name, and then we need to get the number of students
* Set the index to school_name, and then use the count() method for the student_name

In [62]:
per_passing_math_reading = per_passing_math_reading.groupby(["school_name"]).count()["student_name"]

#### E) Calculate the overall passing percentage.
* Calculate the percentage of students who passed math and reading by dividing the the total number of students and multiplying by 100

In [63]:
per_overall_passing_percentage = per_passing_math_reading / per_school_counts * 100
per_overall_passing_percentage

Bailey High School       54.642283
Cabrera High School      91.334769
Figueroa High School     53.204476
Ford High School         54.289887
Griffin High School      90.599455
Hernandez High School    53.527508
Holden High School       89.227166
Huang High School        53.513884
Johnson High School      53.539172
Pena High School         90.540541
Rodriguez High School    52.988247
Shelton High School      89.892107
Thomas High School       90.948012
Wilson High School       90.582567
Wright High School       90.333333
dtype: float64

# 4.8.6 Create the School Summary DataFrame

## 36) We can add the following values to a new DataFrame named per_school_summary_df now that we have performed all the calculations needed for the school summary,

The data and columns of the DataFrame will be:
* Type of school in the "School Type" column
* Total students per school in the "Total Students" column
* Total budget per school in the "Total School Budget" column
* Total budget per student for each school in the "Per Student Budget" column
* Average math score for each school in the "Average Math Score" column
* Average reading score for each school in the "Average Reading Score" column
* Percentage of students passing math for each school in the "% Passing Math" column
* Percentage of students passing reading for each school in the "% Passing Reading" column
* Overall passing percentage for each school in the "% Overall Passing" column

To create per_school_summary_df the column names will be the keys, and the values will each a piece of data we retrieved or calculated. 

### A) Adding a list of values with keys to create a new DataFrame.
Output: The columns are in the correct order, but we still need to apply formatting to the Total School Budget and Per Student Budget columns, as well as apply a thousands separator to the Total Students column. 

In [64]:
per_school_summary_df = pd.DataFrame({
            "School Type": per_school_types,
            "Total Students": per_school_counts,
            "Total School Budget": per_school_budget,
            "Per Student Budget": per_school_capita,
            "Average Math Score": per_school_math,
           "Average Reading Score": per_school_reading,
           "% Passing Math": per_school_passing_math,
           "% Passing Reading": per_school_passing_reading,
           "% Overall Passing": per_overall_passing_percentage})
per_school_summary_df.head()

School Type  Total Students  Total School Budget  \
Bailey High School      District            4976              3124928   
Cabrera High School      Charter            1858              1081356   
Figueroa High School    District            2949              1884411   
Ford High School        District            2739              1763916   
Griffin High School      Charter            1468               917500   

                      Per Student Budget  Average Math Score  \
Bailey High School                 628.0           77.048432   
Cabrera High School                582.0           83.061895   
Figueroa High School               639.0           76.711767   
Ford High School                   644.0           77.102592   
Griffin High School                625.0           83.351499   

                      Average Reading Score  % Passing Math  \
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

# 4.8.7Clean Up the DataFrame (per_school_summary_df)

### 37) In the per_school_summary_df DataFrame, we will format the Total School Budget and Per Student Budget columns to include:
* A U.S. dollar sign 
* Two decimal places
* A thousands separator

#### A) Format the Total School Budget and Per Student Budget columns

In [65]:
per_school_summary_df["Total School Budget"] = per_school_summary_df["Total School Budget"].map("${:,.2f}".format)
per_school_summary_df["Per Student Budget"] = per_school_summary_df["Per Student Budget"].map("${:,.2f}".format)
per_school_summary_df.head()

School Type  Total Students Total School Budget  \
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

# 4.9 High and Low Performing Schools

## 4.9.1 Find the Highest performing schools
To display the top five highest-performing schools, we can sort the per_school_summary_df DataFrame on the % Overall Passing column
* Use df.head() for first 5 rows of df
* Use df.tail() for last 5 rows of df

Using the Pandas sort_values() function, we can sort a DataFrame or Series for a given text, index, or column that is passed within the parentheses. 

In addition, we can add the parameter ascending=False to sort from highest to lowest, or ascending=True to sort from lowest to highest the value referenced in the sort_values() function. 
* If we don't add the ascending parameter, the default is ascending=True.

In [66]:
top_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=False)
top_schools.head()

School Type  Total Students Total School Budget  \
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   
Pena High School        Charter             962         $585,858.00   

                    Per Student Budget  Average Math Score  \
Cabrera High School            $582.00           83.061895   
Thomas High School             $638.00           83.418349   
Griffin High School            $625.00           83.351499   
Wilson High School             $578.00           83.274201   
Pena High School               $609.00           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading  \
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   
Pena High School                 84.044699       94.594595          95.945946   

                     % Overall Passing  
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

## 4.9.2 Find the Lowest Performing Schools
To find the five lowest-performing schools based on the overall percentage of students who passed in the per_school_summary_df DataFrame, we can use the ascending=True parameter. 
* Since ascending=True is the default parameter for the sort_values() function, we don't need to add it. 
* However, it's beneficial to add this parameter so we know how we are sorting

### A) Sort and show top 5 schools

In [67]:
bottom_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=True)
bottom_schools.head()

School Type  Total Students Total School Budget  \
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Hernandez High School    District            4635       $3,022,020.00   
Johnson High School      District            4761       $3,094,650.00   

                      Per Student Budget  Average Math Score  \
Rodriguez High School            $637.00           76.842711   
Figueroa High School             $639.00           76.711767   
Huang High School                $655.00           76.629414   
Hernandez High School            $652.00           77.289752   
Johnson High School              $650.00           77.072464   

                       Average Reading Score  % Passing Math  \
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   

                       % Passing Reading  % Overall Passing  
Rodriguez High School          80.220055          52.988247  
Figueroa High School           80.739234          53.204476  
Huang High School              81.316421          53.513884  
Hernandez High School          80.862999          53.527508  
Johnson High School            81.222432          53.539172

# 4.10 Average Math and Reading Scores by Grade

## 4.10.1 Create Grade-Level DataFrames
You are tasked with creating two new DataFrames: 
* One that will display the average math scores for each grade
* One that will display the average reading scores for each grade. 

These will be grouped by the name of the school.

In the final DataFrame, we need to have: 
* The index on the school name and 
* The average math or reading score for each grade: 9th, 10th, 11th, and 12th. 

### A) Create a grade level DataFrames.

Let's first get the ninth graders in a Series. 
* Set the variable ninth_graders equal to school_data_complete_df["grade"] == "9th". 
This will retrieve all rows where this statement is "True."

Using the code school_data_complete_df["grade"] == "9th", we can filter school_data_complete_df for the "True" cases. 

Now repurpose this code for each grade by replacing "9th" with each grade. 

Assign a grade-level variable to each grade-level DataFrame; for example, ninth_graders is for the ninth grade data.

In [68]:
ninth_graders = school_data_complete_df[(school_data_complete_df["grade"] == "9th")]
tenth_graders = school_data_complete_df[(school_data_complete_df["grade"] == "10th")]
eleventh_graders = school_data_complete_df[(school_data_complete_df["grade"] == "11th")]
twelfth_graders = school_data_complete_df[(school_data_complete_df["grade"] == "12th")]
ninth_graders.head()

Student ID     student_name gender grade        school_name  \
0            0     Paul Bradley      M   9th  Huang High School   
4            4       Bonnie Ray      F   9th  Huang High School   
5            5    Bryan Miranda      M   9th  Huang High School   
12          12  Brittney Walker      F   9th  Huang High School   
13          13     William Long      M   9th  Huang High School   

    reading_score  math_score  School ID      type  size   budget  
0              66          79          0  District  2917  1910635  
4              97          84          0  District  2917  1910635  
5              94          94          0  District  2917  1910635  
12             64          79          0  District  2917  1910635  
13             71          79          0  District  2917  1910635

## 4.10.2 Score Averages Grouped by School Name

### A) Get the Average Math Scores by School
Group each grade level DataFrame by the school name for the average math score.
* For each grade level DataFrame, we will use the groupby() function on the school_name column and apply the mean() on the math_score column

In [69]:
ninth_grade_math_scores = ninth_graders.groupby(["school_name"]).mean()["math_score"]
tenth_grade_math_scores = tenth_graders.groupby(["school_name"]).mean()["math_score"]
eleventh_grade_math_scores = eleventh_graders.groupby(["school_name"]).mean()["math_score"]
twelfth_grade_math_scores = twelfth_graders.groupby(["school_name"]).mean()["math_score"]
eleventh_grade_math_scores

school_name
Bailey High School       77.515588
Cabrera High School      82.765560
Figueroa High School     76.884344
Ford High School         76.918058
Griffin High School      83.842105
Hernandez High School    77.136029
Holden High School       85.000000
Huang High School        76.446602
Johnson High School      77.491653
Pena High School         84.328125
Rodriguez High School    76.395626
Shelton High School      83.383495
Thomas High School       83.498795
Wilson High School       83.195326
Wright High School       83.836782
Name: math_score, dtype: float64

### B) Get the Average Reading Scores by School
Group each grade level DataFrame by the school name for the average reading score.
* Refactor code from previous step

In [70]:
ninth_grade_reading_scores = ninth_graders.groupby(["school_name"]).mean()["reading_score"]
tenth_grade_reading_scores = tenth_graders.groupby(["school_name"]).mean()["reading_score"]
eleventh_grade_reading_scores = eleventh_graders.groupby(["school_name"]).mean()["reading_score"]
twelfth_grade_reading_scores = twelfth_graders.groupby(["school_name"]).mean()["reading_score"]
twelfth_grade_reading_scores

school_name
Bailey High School       80.912451
Cabrera High School      84.287958
Figueroa High School     81.384863
Ford High School         80.662338
Griffin High School      84.013699
Hernandez High School    80.857143
Holden High School       84.698795
Huang High School        80.305983
Johnson High School      81.227564
Pena High School         84.591160
Rodriguez High School    80.376426
Shelton High School      82.781671
Thomas High School       83.831361
Wilson High School       84.317673
Wright High School       84.073171
Name: reading_score, dtype: float64

## 4.10.3 Combine Each Grade Level Series into a Data Frame

### A) Combine each grade level Series for average math scores by school into a single DataFrame
Add the math scores for each grade level to a new DataFrame

In [71]:
math_scores_by_grade = pd.DataFrame({
               "9th": ninth_grade_math_scores,
               "10th": tenth_grade_math_scores,
               "11th": eleventh_grade_math_scores,
               "12th": twelfth_grade_math_scores})
math_scores_by_grade.head()

9th       10th       11th       12th
school_name                                                     
Bailey High School    77.083676  76.996772  77.515588  76.492218
Cabrera High School   83.094697  83.154506  82.765560  83.277487
Figueroa High School  76.403037  76.539974  76.884344  77.151369
Ford High School      77.361345  77.672316  76.918058  76.179963
Griffin High School   82.044010  84.229064  83.842105  83.356164

### B) Combine each grade level Series for average reading scores by school into a single DataFrame.
Use the same format to add the reading scores for each grade level to a new DataFrame

In [72]:
reading_scores_by_grade = pd.DataFrame({
              "9th": ninth_grade_reading_scores,
              "10th": tenth_grade_reading_scores,
              "11th": eleventh_grade_reading_scores,
              "12th": twelfth_grade_reading_scores})
reading_scores_by_grade.head()

9th       10th       11th       12th
school_name                                                     
Bailey High School    81.303155  80.907183  80.945643  80.912451
Cabrera High School   83.676136  84.253219  83.788382  84.287958
Figueroa High School  81.198598  81.408912  80.640339  81.384863
Ford High School      80.632653  81.262712  80.403642  80.662338
Griffin High School   83.369193  83.706897  84.288089  84.013699

## 4.10.4 Format the Averages and Remove the Index Name
* For reporting purposes, we'll format the grade-level averages to one decimal place, as well as remove the name of the index column, school_name.
* To format the reading and math averages for each grade level we'll use map("{:.1f}".format. 
* We'll remove the index, school_name, by setting the index name to "None" with the following syntax: index.name = None.

### A) Format Each Column (math_scores_by_grade)

In [73]:
math_scores_by_grade["9th"] = math_scores_by_grade["9th"].map("{:.1f}".format)
math_scores_by_grade["10th"] = math_scores_by_grade["10th"].map("{:.1f}".format)
math_scores_by_grade["11th"] = math_scores_by_grade["11th"].map("{:.1f}".format)
math_scores_by_grade["12th"] = math_scores_by_grade["12th"].map("{:.1f}".format)

### B) Make sure the columns are in the correct order

In [74]:
math_scores_by_grade = math_scores_by_grade[
                 ["9th", "10th", "11th", "12th"]]

### C) Remove the Index Name

In [75]:
math_scores_by_grade.index.name = None

### D) Display the dataFrame

In [76]:
math_scores_by_grade.head()

9th  10th  11th  12th
Bailey High School    77.1  77.0  77.5  76.5
Cabrera High School   83.1  83.2  82.8  83.3
Figueroa High School  76.4  76.5  76.9  77.2
Ford High School      77.4  77.7  76.9  76.2
Griffin High School   82.0  84.2  83.8  83.4

### E) Format Each Column (reading_scores_by_grade)

In [77]:
reading_scores_by_grade["9th"] = reading_scores_by_grade["9th"].map("{:,.1f}".format)
reading_scores_by_grade["10th"] = reading_scores_by_grade["10th"].map("{:,.1f}".format)
reading_scores_by_grade["11th"] = reading_scores_by_grade["11th"].map("{:,.1f}".format)
reading_scores_by_grade["12th"] = reading_scores_by_grade["12th"].map("{:,.1f}".format)

### F) Make sure the columns are in the correct order

In [78]:
reading_scores_by_grade = reading_scores_by_grade[
                 ["9th", "10th", "11th", "12th"]]

### G) Remove the Index Name

In [79]:
reading_scores_by_grade.index.name = None

### H) Display the dataFrame

In [80]:
reading_scores_by_grade.head()

9th  10th  11th  12th
Bailey High School    81.3  80.9  80.9  80.9
Cabrera High School   83.7  84.3  83.8  84.3
Figueroa High School  81.2  81.4  80.6  81.4
Ford High School      80.6  81.3  80.4  80.7
Griffin High School   83.4  83.7  84.3  84.0

# 4.11 Group Scores by School Spending per Student

## 4.11.1 Establish the Spending Ranges per Student
* For this analysis, we will need to sort the school spending per student into four spending "bins," or ranges. 
* We will need to arbitrarily determine the spending ranges in order to group the schools fairly. 
* Each bin must include the average math and reading scores, the percentage passing math and reading, and the overall passing percentage for the spending bins, not for the schools.
* The final DataFrame will have the spending bins included in place of the box that says "Spending Bins."  (We'll determine the four spending bins later in this section.) 

To create this DataFrame, we need to do the following:
1.	Group the schools into four bins, or ranges, based on the budget per student.
2.	For each spending range, get the following data:
* Average math and reading scores
* The percentage of students passing math and reading
* The overall passing percentage, which is the average of the percentage of students passing math and reading

Before we aggregate the school spending per student into four bins, we must establish the spending ranges. 
* We can determine the spending ranges by using the per_school_capita Series that was used to get the budget per student.

### A) Get the Spending Ranges
Get the descriptive statistics for the per_school_capita.

Output: 
* As you can see, the minimum is 578 and the maximum is 655. 
* The standard deviation is 28.5, or approximately 30. 
* We don't want the lowest bin to be 578 because there is only one school at or below 578, which is Wilson High School. 
* However, there are four schools that spend less than 585 per student, so 585 will be our lowest bin. 
* Also, because the standard deviation is about 30, we will increase the bins by 30, up to 675. 
* Therefore, the four bins will be: 585, 615, 645, and 675.

Let's go over why these spending bins have five numbers instead of four, which is the number of bins we want. 
* When Pandas creates the lower edge, it needs a value lower than the lowest value in the column of the per_school_capita Series, which happens to be 578. 

In our DataFrame, we make the lower edge equal to 0. 
* If we don't include the 0 for the spending bins, the lowest bin becomes 585–614, which means that the schools that spend lessthan 585 are not considered. 
* First we'll use the correct spending bins for our ranges, and then we'll leave out the 0 to see how that affects the bins.

Let's address one more item before moving on. 
* If you look at the spending_bins list, you'll notice that the dollar sign  isn't included with the ranges. 
* This is because the numbers in the  per_school_capita Series are the float64 data type, so adding a dollar sign would cause an error.

In [81]:
per_school_capita.describe()

count     15.000000
mean     620.066667
std       28.544368
min      578.000000
25%      591.500000
50%      628.000000
75%      641.500000
max      655.000000
dtype: float64

### B) Group the Series on the Spending Ranges
Cut the per_school_capita into the spending ranges.


What happens if we don't add 0 as the lower edge for the first bin? 

spending_bins = [585, 615, 645, 675]

* Output: Four high schools—Cabrera, Holden, Wilson, and Wright—are not in a bin and therefore would not be in the final DataFrame.

spending_bins = [0, 585, 615, 645, 675]
pd.cut(per_school_capita, spending_bins)

* Output: Each school has a range for the spending per student instead of a specific amount


spending_bins = [0, 585, 615, 645, 675]
per_school_capita.groupby(pd.cut(per_school_capita, spending_bins)).count()

* Output: Number of schools in each range (0-585 = 4; 586-615 = 2; 615-645 = 6; 645-675 = 3) 
* Looking at this output, it seems that we didn't group the schools fairly. 

We need to adjust our ranges so that we have three or four schools in each range. 
* Let's increase the [585, 615] range to [585, 630] and change the third range to [630, 645]. 

The spending_bins will look like this: 
[0, 585, 630, 645, 675]

spending_bins = [0, 585, 630, 645, 675]
per_school_capita.groupby(pd.cut(per_school_capita, spending_bins)).count()

Output: Number of schools in each range (0-585 = 4; 586-615 = 4; 615-645 = 4; 645-675 = 3)

In [82]:
spending_bins = [0, 585, 630, 645, 675]
per_school_capita.groupby(pd.cut(per_school_capita, spending_bins)).count()

(0, 585]      4
(585, 630]    4
(630, 645]    4
(645, 675]    3
dtype: int64

In [83]:
spending_bins = [0, 585, 630, 645, 675]
group_names = ["<$584", "$585-629", "$630-644", "$645-675"]

## 4.11.2 Categorize the Spending Bins
Using our spending bins and ranges, we can create a new column in the per_school_summary_df DataFrame which will be assigned the spending bins from the per_school_capita Series.

To do this, we will need to do the following:
* Use the cut() function on the per_school_capita Series.
* Add the bin data to a new column in the per_school_summary_df DataFrame.

There's a lot going on in this code, so let's break it down.
* To the left of the equals sign, we add a new column called "Spending Ranges (Per Student)"{style="font-size: 1rem;"} to the per_school_summary_df DataFrame. This column will contain the values specified by using the cut() function.
* Inside the cut() function, we add the data (per_school_capita) we are going to use for "Spending Ranges (Per Student)". The data must be a one-dimensional array, like a list or the per_school_capita Series.
* Inside the parentheses, we add the spending_bins and the labels for the bins using labels=group_names.

In [84]:
per_school_summary_df["Spending Ranges (Per Student)"] = pd.cut(per_school_capita, spending_bins, labels=group_names)
per_school_summary_df.head()

School Type  Total Students Total School Budget  \
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  \
Bailey High School            81.933280          54.642283   
Cabrera High School           97.039828          91.334769   
Figueroa High School          80.739234          53.204476   
Ford High School              79.299014          54.289887   
Griffin High School           97.138965          90.599455   

                     Spending Ranges (Per Student)  
Bailey High School                        $585-629  
Cabrera High School                          <$584  
Figueroa High School                      $630-644  
Ford High School                          $630-644  
Griffin High School                       $585-629

## 4.11.3 Group by the Spending Ranges
Calculate averages for the desired columns.

In [85]:
spending_math_scores = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Math Score"]
spending_reading_scores = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Reading Score"]

spending_passing_math = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Math"]
spending_passing_reading = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Reading"]

overall_passing_spending = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Overall Passing"]

overall_passing_spending

Spending Ranges (Per Student)
<$584       90.369459
$585-629    81.418596
$630-644    62.857656
$645-675    53.526855
Name: % Overall Passing, dtype: float64

## 4.11.4 Create a DataFrame for the Scores by School Spending

### A) Create the Spending Summary DataFrame
Assemble into DataFrame

In [86]:
spending_summary_df = pd.DataFrame({
          "Average Math Score" : spending_math_scores,
          "Average Reading Score": spending_reading_scores,
          "% Passing Math": spending_passing_math,
          "% Passing Reading": spending_passing_reading,
          "% Overall Passing": overall_passing_spending})
spending_summary_df

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$584                                   83.455399              83.933814   
$585-629                                81.899826              83.155286   
$630-644                                78.518855              81.624473   
$645-675                                76.997210              81.027843   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$584                               93.460096          96.610877   
$585-629                            87.133538          92.718205   
$630-644                            73.484209          84.391793   
$645-675                            66.164813          81.133951   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$584                                  90.369459  
$585-629                               81.418596  
$630-644                               62.857656  
$645-675                               53.526855

### B) Format the DataFrame
Formatting

In [87]:
spending_summary_df["Average Math Score"] = spending_summary_df["Average Math Score"].map("{:.1f}".format)
spending_summary_df["Average Reading Score"] = spending_summary_df["Average Reading Score"].map("{:.1f}".format)

spending_summary_df["% Passing Math"] = spending_summary_df["% Passing Math"].map("{:.0f}".format)
spending_summary_df["% Passing Reading"] = spending_summary_df["% Passing Reading"].map("{:.0f}".format)

spending_summary_df["% Overall Passing"] = spending_summary_df["% Overall Passing"].map("{:.0f}".format)

spending_summary_df

Average Math Score Average Reading Score  \
Spending Ranges (Per Student)                                            
<$584                                       83.5                  83.9   
$585-629                                    81.9                  83.2   
$630-644                                    78.5                  81.6   
$645-675                                    77.0                  81.0   

                              % Passing Math % Passing Reading  \
Spending Ranges (Per Student)                                    
<$584                                     93                97   
$585-629                                  87                93   
$630-644                                  73                84   
$645-675                                  66                81   

                              % Overall Passing  
Spending Ranges (Per Student)                    
<$584                                        90  
$585-629                                     81  
$630-644                                     63  
$645-675                                     54

# 4.12. Group Scores by School Size

## 4.12.1 Create Bins for School Size
Establish the bins.

Looking at the per_school_counts Series, we can see that the highest student population is 4,976. 
* Therefore, the upper edge will be 5,000. 

Because we have only three bins—small, medium, and large—here's how we can group the school sizes:
* "Small" schools have fewer than 1,000 students
* "Medium" schools have 1,000–1,999 students
* "Large" schools have 2,000–5,000 students

In [88]:
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

## 4.12.2 Categorize the Size Bins
Categorize spending based on the bins.

In [89]:
per_school_summary_df["School Size"] = pd.cut(per_school_summary_df["Total Students"], size_bins, labels=group_names)
per_school_summary_df.head()

School Type  Total Students Total School Budget  \
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  \
Bailey High School            81.933280          54.642283   
Cabrera High School           97.039828          91.334769   
Figueroa High School          80.739234          53.204476   
Ford High School              79.299014          54.289887   
Griffin High School           97.138965          90.599455   

                     Spending Ranges (Per Student)         School Size  
Bailey High School                        $585-629   Large (2000-5000)  
Cabrera High School                          <$584  Medium (1000-2000)  
Figueroa High School                      $630-644   Large (2000-5000)  
Ford High School                          $630-644   Large (2000-5000)  
Griffin High School                       $585-629  Medium (1000-2000)

## 4.12.3 Group by School Size
Calculate averages for the desired columns.

The final DataFrame will contain the average math and reading scores, the percentage of students who passed math and reading, and the overall percentage of students who passed for each school-size bin. 

Just as we did when we created the DataFrame for school spending per student, we'll use the groupby() function to group the data on the size bins in the index column, School Size.

First, let's create four new Series. We need to get the following data:
* Average Math Score
* Average Reading Score
* Average % Passing Math
* Average % Passing Reading
* Average % Overall Passing

Create each Series using the groupby() function on the School Size column. 

The data for each column will be calculated by using the mean() method to get the averages of the following columns: Average Math Score, Average Reading Score, % Passing Math, and % Passing Reading. 

For the average "% Overall Passing," we'll add the % Passing Math and  % Passing Reading columns, and then divide by 2.

In [90]:
size_math_scores = per_school_summary_df.groupby(["School Size"]).mean()["Average Math Score"]
size_reading_scores = per_school_summary_df.groupby(["School Size"]).mean()["Average Reading Score"]

size_passing_math = per_school_summary_df.groupby(["School Size"]).mean()["% Passing Math"]
size_passing_reading = per_school_summary_df.groupby(["School Size"]).mean()["% Passing Reading"]

size_overall_passing = per_school_summary_df.groupby(["School Size"]).mean()["% Overall Passing"]

## 4.12.4 Create a DataFrame for the Scores by School Size

#### A ) Creat e the size_summary dataframe
Assemble into DataFrame

Just like we did with the school spending data, we are going to add the average math and reading scores, the average percentage of students who passed math and reading, and the average overall percentage to a new DataFrame.

In [91]:
size_summary_df = pd.DataFrame({
          "Average Math Score" : size_math_scores,
          "Average Reading Score": size_reading_scores,
          "% Passing Math": size_passing_math,
          "% Passing Reading": size_passing_reading,
          "% Overall Passing": size_overall_passing})
size_summary_df

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)               96.099437          89.883853  
Medium (1000-2000)          96.790680          90.621535  
Large (2000-5000)           82.766634          58.286003

#### B) Format the DataFrame
Formatting

In [92]:
size_summary_df["Average Math Score"] = size_summary_df["Average Math Score"].map("{:.1f}".format)
size_summary_df["Average Reading Score"] = size_summary_df["Average Reading Score"].map("{:.1f}".format)

size_summary_df["% Passing Math"] = size_summary_df["% Passing Math"].map("{:.0f}".format)
size_summary_df["% Passing Reading"] = size_summary_df["% Passing Reading"].map("{:.0f}".format)

size_summary_df["% Overall Passing"] = size_summary_df["% Overall Passing"].map("{:.0f}".format)

size_summary_df

Average Math Score Average Reading Score % Passing Math  \
School Size                                                                  
Small (<1000)                    83.8                  83.9             94   
Medium (1000-2000)               83.4                  83.9             94   
Large (2000-5000)                77.7                  81.3             70   

                   % Passing Reading % Overall Passing  
School Size                                             
Small (<1000)                     96                90  
Medium (1000-2000)                97                91  
Large (2000-5000)                 83                58

# 4.13 Group Series by School Type

## 4.13.1 Group by school type
Calculate averages for the desired columns.

In [95]:
type_math_scores = per_school_summary_df.groupby(["School Type"]).mean()["Average Math Score"]
type_reading_scores = per_school_summary_df.groupby(["School Type"]).mean()["Average Reading Score"]

type_passing_math = per_school_summary_df.groupby(["School Type"]).mean()["% Passing Math"]
type_passing_reading = per_school_summary_df.groupby(["School Type"]).mean()["% Passing Reading"]

type_overall_passing = per_school_summary_df.groupby(["School Type"]).mean()["% Overall Passing"]

## 4.13.2 Create a DataFrame for the Scores by School Type

### A) Create the School Type DataFrame
Assemble into DataFrame

In [96]:
type_summary_df = pd.DataFrame({
          "Average Math Score" : type_math_scores,
          "Average Reading Score": type_reading_scores,
          "% Passing Math": type_passing_math,
          "% Passing Reading": type_passing_reading,
          "% Overall Passing": type_overall_passing})
type_summary_df

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter              96.586489          90.432244  
District             80.799062          53.672208

### B) Format the DataFrame
Formatting

In [97]:
type_summary_df["Average Math Score"] = type_summary_df["Average Math Score"].map("{:.1f}".format)
type_summary_df["Average Reading Score"] = type_summary_df["Average Reading Score"].map("{:.1f}".format)

type_summary_df["% Passing Math"] = type_summary_df["% Passing Math"].map("{:.0f}".format)
type_summary_df["% Passing Reading"] = type_summary_df["% Passing Reading"].map("{:.0f}".format)

type_summary_df["% Overall Passing"] = type_summary_df["% Overall Passing"].map("{:.0f}".format)

type_summary_df

Average Math Score Average Reading Score % Passing Math  \
School Type                                                           
Charter                   83.5                  83.9             94   
District                  77.0                  81.0             67   

            % Passing Reading % Overall Passing  
School Type                                      
Charter                    97                90  
District                   81                54